File: api/src/main/java/moe/caa/multilogin/api/logger/Logger.java
package moe.caa.multilogin.api.logger;

import org.jetbrains.annotations.ApiStatus;

@ApiStatus.Internal
public abstract class Logger {
    private void handleLog0(Level level, String message, Throwable throwable){
        if (level == Level.DEBUG && LoggerProvider.debugMode) {
            level = Level.INFO;
        }
        handleLog(level, message, throwable);
    }

    protected abstract void handleLog(Level level, String message, Throwable throwable);
    public void log(Level level, String message, Throwable throwable){handleLog0(level, message, throwable);}
    public void log(Level level, String message) {handleLog0(level, message, null);}
    public void log(Level level, Throwable throwable) {handleLog0(level, null, throwable);}
    public void debug(String message, Throwable throwable) {handleLog0(Level.DEBUG, message, throwable);}
    public void debug(String message) {log(Level.DEBUG, message);}
    public void debug(Throwable throwable) {handleLog0(Level.DEBUG, null, throwable);}
    public void info(String message, Throwable throwable) {handleLog0(Level.INFO, message, throwable);}
    public void info(String message) {log(Level.INFO, message);}
    public void info(Throwable throwable) {handleLog0(Level.INFO, null, throwable);}
    public void warn(String message, Throwable throwable) {handleLog0(Level.WARN, message, throwable);}
    public void warn(String message) {log(Level.WARN, message);}
    public void warn(Throwable throwable) {handleLog0(Level.WARN, null, throwable);}
    public void error(String message, Throwable throwable) {handleLog0(Level.ERROR, message, throwable);}
    public void error(String message) {log(Level.ERROR, message);}
    public void error(Throwable throwable) {handleLog0(Level.ERROR, null, throwable);}
}

File: api/src/main/kotlin/moe/caa/multilogin/api/logger/Logger.kt
package moe.caa.multilogin.api.logger

import moe.caa.multilogin.api.logger.bridge.ConsoleLogger
import moe.caa.multilogin.api.logger.bridge.DebugLogger
import org.jetbrains.annotations.ApiStatus.Internal

@Internal
interface Logger {
    companion object {
        var logger: Logger = ConsoleLogger()

        fun debug(debug: Boolean) {
            if (debug) {
                if (logger !is DebugLogger) {
                    logger = DebugLogger(logger)
                    logger.warn("Debug mode on.")
                }
            } else {
                if (logger is DebugLogger) {
                    logger = (logger as DebugLogger).originLogger
                    logger.warn("Debug mode off.")
                    debug(false)
                }
            }
        }
    }

    fun log(level: Level, message: String? = null, throwable: Throwable? = null)
    fun debug(message: String? = null, throwable: Throwable? = null) = log(Level.DEBUG, message, throwable)
    fun info(message: String? = null, throwable: Throwable? = null) = log(Level.INFO, message, throwable)
    fun warn(message: String? = null, throwable: Throwable? = null) = log(Level.WARN, message, throwable)
    fun error(message: String? = null, throwable: Throwable? = null) = log(Level.ERROR, message, throwable)
}

fun log(level: Level, message: String? = null, throwable: Throwable? = null) =
    Logger.logger.log(level, message, throwable)

fun logInfo(message: String? = null, throwable: Throwable? = null) = Logger.logger.info(message, throwable)
fun logDebug(message: String? = null, throwable: Throwable? = null) = Logger.logger.debug(message, throwable)
fun logWarn(message: String? = null, throwable: Throwable? = null) = Logger.logger.warn(message, throwable)
fun logError(message: String? = null, throwable: Throwable? = null) = Logger.logger.error(message, throwable)